### ISSUE TITLE
Existing window is not focused when on another workspace and URL is opened

### ISSUE THREAD

**Issue by abbradar** [The-Compiler/qutebrowser#1866]
Configuration:

* OS: NixOS 16.09.git.09f67a7 (Flounder)
* WM: xmonad
* qutebrowser: 0.8.2
* general.new-instance-open-target: tab

When qutebrowser window is on not currently selected workspace, opening a new URL (e.g. with `qutebrowser https://ddg.gg`) won't switch workspace to the one with the browser and focus the window.

**Comment by The-Compiler** [comment]
@haasn rightly points out that this should be configurable:

```
21:23 <haasn> The-Compiler: re: qutebrowser's behavior on new tab, if you add more “activation methods” then they should be configurable. I'm very happy with the default of just showing a hint
21:23 <haasn> an urgent hint*
21:24 <haasn> firefox's “steal focus” anti-feature forced me to use the equivalent of tab-bg in firefox, but I don't really want that
```

**Comment by lahwaacz** [comment]
It's not the browser's task to activate other windows, let alone change workspaces. It can only give hints to the window manager (via [_NET_ACTIVE_WINDOW](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472702304)), which can either oblige or refuse the request.

**Comment by abbradar** [comment]
Yeah, but I presumed that qutebrowser doesn't send that hint... I can look at how Firefox does that later (it does switch the workspace).

**Comment by abbradar** [comment]
I'm still not sure but I think it's because Firefox calls `XRaiseWindow` instead of sending the hint. At least it's hinted in [a bug report](https://bugzilla.mozilla.org/show_bug.cgi?id=392721#c0) and the code contains `XRaiseWindow`, too.

**Comment by Kingdread** [comment]
Qt also sends the `_NET_ACTIVE_WINDOW` (which if handled by the WM should change workspaces) message, but only if the WM supports it (in the `_NET_SUPPORTED` property). Otherwise, it just uses `XSetInputFocus` (or rather the xcb equivalent, `xcb_set_input_focus`), which won't change workspaces.

qutebrowser actually uses this method if your target is not `tab-silent`, so the requested behavior *should* already work (if your WM supports it, as @lahwaacz points out).

You're using xmonad, and if I remember correctly, EWMH support has to be enabled manually in the configuration for xmonad. Could you please check if that is the case on your machine? Or could you check whether you have any restrictions set, i.e. that not every program is allowed to steal focus?

**Comment by abbradar** [comment]
@Kingdread Yes, I use xmonad and EWMH support is enabled. I have `_NET_ACTIVE_WINDOW` in `_NET_SUPPORTED`.

From reading documentation, qutebrowser needs to open needed desktop by itself after all -- for example, that's how xdotool does it: https://github.com/jordansissel/xdotool/blob/master/xdo.c#L459 (it works for me).

I'm not sure however if that's possible without resorting to non-portable code...

**Comment by abbradar** [comment]
To give a quick overview of what exactly xdotool does:
1. Finds out the window's workspace via `_NET_WM_DESKTOP`;
2. Sets current workspace via `_NET_CURRENT_DESKTOP`.

**Comment by lahwaacz** [comment]
> I have `_NET_ACTIVE_WINDOW` in `_NET_SUPPORTED`.

That still does not mean that the WM _has to_ switch the desktop on every request. From [the standard](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472702304):

> Depending on the information provided with the message, the Window Manager may decide to refuse the request (either completely ignore it, or e.g. use _NET_WM_STATE_DEMANDS_ATTENTION).

That still does not say that the request should be handled at all. Depending on the WM, the behaviour can be anything. I have no idea how xmonad handles it, but e.g. i3 has [4 options](https://github.com/i3/i3/blob/4bec3b9d24fe0bd7bab4792497bbd02bffaa6620/include/config.h#L172-L182) for this (see also the [user guide](http://i3wm.org/docs/userguide.html#focus_on_window_activation)).

**Comment by Kingdread** [comment]
Seems like XMonad honors every request: https://github.com/xmonad/xmonad-contrib/blob/master/XMonad/Hooks/EwmhDesktops.hs#L131-L165 (though I'm no XMonad user and a Haskell noob).

I'll maybe test it myself sometime with a little application that only uses Qt's `activateWindow`, but it might be that XMonad expects us to switch the desktop first, in which case yeah, we probably can't do that with non-portable code...

**Comment by abbradar** [comment]
> That still does not mean that the WM has to switch the desktop on every request.

Of course. To clarify: xmonad does switch the focus when qutebrowser is on the same workspace I open the link from.

> I have no idea how xmonad handles it, but e.g. i3 has 4 options for this (see also the user guide).

Unfortunately, xmonad doesn't have much settings on this front -- we can just enable event handling and that's all (looking through source code).

**Comment by tbsmn** [comment]
This ticket is a bit older, but I have the same problem with MacOS (10.14.6) and qutebrowser v1.10.1. When I click on an external link the space does not switch to the space qutebrowser is currently active on.

I have set `new_instance_open_target` to `tab` in the configuration.

Is there a solution for this problem?
